app/src/gecko/java/org/mozilla/focus/web/GeckoViewPrompt.java

```diff
@@ -0,0 +1,888 @@
+                            // Clear
+                            callback.confirm("");
+                            return;
+                        }
+                        if (datePicker != null) {
+                            cal.set(datePicker.getYear(), datePicker.getMonth(),
+                                    datePicker.getDayOfMonth());
+                        }
+                        if (timePicker != null) {
+                            setCalendarTime(cal, timePicker);
+                        }
+                        callback.confirm(formatter.format(cal.getTime()));
+                    }
+                };
+        builder.setNegativeButton(android.R.string.cancel, /* listener */ null)
+                .setNeutralButton(R.string.gv_prompt_clear, listener)
+                .setPositiveButton(android.R.string.ok, listener);
+        createStandardDialog(builder, callback).show();
+    }
+
+    public void promptForFile(GeckoSession session, String title, int type,
+                              String[] mimeTypes, FileCallback callback) {
+        final Activity activity = mActivity;
+        if (activity == null) {
+            callback.dismiss();
+            return;
+        }
+
+        // Merge all given MIME types into one, using wildcard if needed.
+        String mimeType = null;
+        String mimeSubtype = null;
+        for (final String rawType : mimeTypes) {
+            final String normalizedType = rawType.trim().toLowerCase(Locale.ROOT);
+            final int len = normalizedType.length();
+            int slash = normalizedType.indexOf('/');
+            if (slash < 0) {
+                slash = len;
+            }
+            final String newType = normalizedType.substring(0, slash);
+            final String newSubtype = normalizedType.substring(Math.min(slash + 1, len));
+            if (mimeType == null) {
+                mimeType = newType;
+            } else if (!mimeType.equals(newType)) {
+                mimeType = "*";
+            }
+            if (mimeSubtype == null) {
+                mimeSubtype = newSubtype;
+            } else if (!mimeSubtype.equals(newSubtype)) {
+                mimeSubtype = "*";
+            }
+        }
+
+        final Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
+        intent.setType((mimeType != null ? mimeType : "*") + '/' +
+                (mimeSubtype != null ? mimeSubtype : "*"));
+        intent.addCategory(Intent.CATEGORY_OPENABLE);
+        intent.putExtra(Intent.EXTRA_LOCAL_ONLY, true);
+        intent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true);
+        intent.putExtra(Intent.EXTRA_MIME_TYPES, mimeTypes);
+
+        try {
+            mFileType = type;
+            mFileCallback = callback;
+            activity.startActivityForResult(intent, filePickerRequestCode);
+        } catch (final ActivityNotFoundException e) {
+            Log.e(LOGTAG, "Cannot launch activity", e);
+            callback.dismiss();
```

Comment:
\[PMD\] INFO: The method 'onFileCallbackResult' has a Cyclomatic Complexity of 10\.